### PR TITLE
initial implementation of a web view in vscode #5

### PR DESCRIPTION
**Pull request by lhein**
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

### CODE REVIEW

**Comment by lhein** [comment]
from the functional pov the extension is working and I think this PR should be ready to merge as is. Further improvements should be done in separate PRs

File: package.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		"compile": "tsc -p ./",
+		"watch": "tsc -w -p ./",
+		"postinstall": "node ./node_modules/vscode/bin/install"
+	},
+	"dependencies": {
+		"fetch": "^1.1.0",
+		"vscode": "^1.1.18"
+	},
+	"devDependencies": {
+		"@types/node": "^10.5.2",
```
Comment (apupier) [inline package.json]: maybe worth using the latest dependencies to start?
10.12.12 exists

File: package.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		"tslint": "^5.11.0",
+		"typescript": "^2.9.2"
```
Comment (apupier) [inline package.json]: maybe worth using the latest dependencies to start?
3.2.1 exists

File: package.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	},
+	"scripts": {
+		"vscode:prepublish": "tsc -p ./",
+		"compile": "tsc -p ./",
+		"watch": "tsc -w -p ./",
+		"postinstall": "node ./node_modules/vscode/bin/install"
+	},
+	"dependencies": {
+		"fetch": "^1.1.0",
+		"vscode": "^1.1.18"
```
Comment (apupier) [inline package.json]: maybe worth using the latest dependencies to start?
1.1.24 exists

File: package.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
+{
+	"name": "atlasmap-viewer",
+	"description": "Opens a browser view for Atlas Map",
+	"version": "0.0.1",
+	"publisher": "lhein",
+	"engines": {
+		"vscode": "^1.25.0"
```
Comment (apupier) [inline package.json]: maybe worth using the latest dependencies to start?
1.29.1 exists

File: src/extension.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			async deserializeWebviewPanel(webviewPanel: vscode.WebviewPanel, state: any) {
+				console.log(`Got state: ${state}`);
+				AtlasMapPanel.revive(webviewPanel, context.extensionPath);
+			}
+		});
+	}
+}
+
+/**
+ * Manages cat coding webview panels
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/extension.ts]: copy paste is my enemy :D

File: src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ */
+class AtlasMapPanel {
+	// the URL of the local atlasmap ui
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/extension.ts]: no its not. this is a initial version of the web view taking any URL we define here. The syndesis staging url was put there just for test purposes.

**Comment by apupier** [comment]
by specifying the base href, at least it seems that it is starting to load them:
image

to test I hardcoded it:

```
AtlasMapPanel.currentPanel._panel.webview.html = 
				`
				<!doctype html>
				<html lang="en">
				
				<head>
				  <meta charset="utf-8">
				  <title>AtlasMap Data Mapper UI</title>
				  <base href="http://127.0.0.1:8585/">
				
				  <meta name="viewport" content="width=device-width, initial-scale=1">
				  <link rel="icon" type="image/x-icon" href="favicon.ico">
				</head>
				
				<body>
				  <div class='atlasmapNavbar'>
					<atlasmap-navbar></atlasmap-navbar>
				  </div>
				
				  <div class="atlasmap__root">
					<atlasmap-dev-root></atlasmap-dev-root> 
				  </div>
				<script type="text/javascript" src="runtime.js"></script><script type="text/javascript" src="polyfills.js"></script><script type="text/javascript" src="styles.js"></script><script type="text/javascript" src="scripts.js"></script><script type="text/javascript" src="vendor.js"></script><script type="text/javascript" src="main.js"></script></body>
				
				</html>
				`;
```

**Comment by apupier** [comment]
there is an "unhandled Navigation error" in the console log:
image

in network, it seems that all calls are in success:
image

**Comment by apupier** [comment]
tried to dig into the stacktrace of the log, the best that I reached is the first place where an error is caught

image

it has a slightly better error message: "No elements in the sequence" but I don't know which sequence it is. I found no specific things to Atlasmap in the stack.

**Comment by apupier** [comment]
if i show "verbose" Console log, there are also the error [Violation] Avoid using document;write().

image

it is thrown by the electron browser: "debugger:///VM93 C:\Users\Aurelien Pupier\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\parts\webview\electron-browser\webview-pre.js:formatted"
is Atlasmap working inside Electron?

**Comment by apupier** [comment]
if i try to open the hardcoded html inside Google chrome, it is well displayed even if the navigation error is also present:
image

**Comment by apupier** [comment]
I tried to use [Min](https://minbrowser.github.io/min/), a minimal web browser based on electron.
using the direct URl or the local html file with hardcoded base href, it is working fine. (but when I try to go back in url field of Min, it seems that it has resolved the url to the base href provided and the local file url is no more provided

image

**Comment by apupier** [comment]
I put a breakpoint on mapping-management.service.ts L;161 (in a catch error which is swallowing the error).

```
error: any
:
DOMException: Failed to read the 'cookie' property from 'Document': Cookies are disabled inside 'data:' URLs. at HttpXsrfCookieExtractor.push../node_modules/@angular/common/fesm5/http.js.HttpXsrfCookieExtractor.getToken (http://127.0.0.1:8585/vendor.js:7688:37) at ApiXsrfInterceptor.push../src/app/lib/atlasmap-data-mapper/services/api-xsrf-interceptor.service.ts.ApiXsrfInterceptor.intercept (http://127.0.0.1:8585/main.js:8780:45) at HttpInterceptorHandler.push../node_modules/@angular/common/fesm5/http.js.HttpInterceptorHandler.handle (http://127.0.0.1:8585/vendor.js:7129:33) at HttpXsrfInterceptor.push../node_modules/@angular/common/fesm5/http.js.HttpXsrfInterceptor.intercept (http://127.0.0.1:8585/vendor.js:7720:25) at HttpInterceptorHandler.push../node_modules/@angular/common/fesm5/http.js.HttpInterceptorHandler.handle (http://127.0.0.1:8585/vendor.js:7129:33) at HttpInterceptingHandler.push../node_modules/@angular/common/fesm5/http.js.HttpInterceptingHandler.handle (http://127.0.0.1:8585/vendor.js:7764:27) at MergeMapSubscriber.project (http://127.0.0.1:8585/vendor.js:6969:184) at MergeMapSubscriber.push../node_modules/rxjs/_esm5/internal/operators/mergeMap.js.MergeMapSubscriber._tryNext (http://127.0.0.1:8585/vendor.js:106781:27) at MergeMapSubscriber.push../node_modules/rxjs/_esm5/internal/operators/mergeMap.js.MergeMapSubscriber._next (http://127.0.0.1:8585/vendor.js:106771:18) at MergeMapSubscriber.push../node_modules/rxjs/_esm5/internal/Subscriber.js.Subscriber.next (http://127.0.0.1:8585/vendor.js:101228:18)
code
:
18
message
:
"Failed to read the 'cookie' property from 'Document': Cookies are disabled inside 'data:' URLs."
name
:
"SecurityError"
stack
:
"Error: Failed to read the 'cookie' property from 'Document': Cookies are disabled inside 'data:' URLs.↵    at HttpXsrfCookieExtractor.push../node_modules/@angular/common/fesm5/http.js.HttpXsrfCookieExtractor.getToken (http://127.0.0.1:8585/vendor.js:7688:37)↵    at ApiXsrfInterceptor.push../src/app/lib/atlasmap-data-mapper/services/api-xsrf-interceptor.service.ts.ApiXsrfInterceptor.intercept (http://127.0.0.1:8585/main.js:8780:45)↵    at HttpInterceptorHandler.push../node_modules/@angular/common/fesm5/http.js.HttpInterceptorHandler.handle (http://127.0.0.1:8585/vendor.js:7129:33)↵    at HttpXsrfInterceptor.push../node_modules/@angular/common/fesm5/http.js.HttpXsrfInterceptor.intercept (http://127.0.0.1:8585/vendor.js:7720:25)↵    at HttpInterceptorHandler.push../node_modules/@angular/common/fesm5/http.js.HttpInterceptorHandler.handle (http://127.0.0.1:8585/vendor.js:7129:33)↵    at HttpInterceptingHandler.push../node_modules/@angular/common/fesm5/http.js.HttpInterceptingHandler.handle (http://127.0.0.1:8585/vendor.js:7764:27)↵    at MergeMapSubscriber.project (http://127.0.0.1:8585/vendor.js:6969:184)↵    at MergeMapSubscriber.push../node_modules/rxjs/_esm5/internal/operators/mergeMap.js.MergeMapSubscriber._tryNext (http://127.0.0.1:8585/vendor.js:106781:27)↵    at MergeMapSubscriber.push../node_modules/rxjs/_esm5/internal/operators/mergeMap.js.MergeMapSubscriber._next (http://127.0.0.1:8585/vendor.js:106771:18)↵    at MergeMapSubscriber.push../node_modules/rxjs/_esm5/internal/Subscriber.js.Subscriber.next (http://127.0.0.1:8585/vendor.js:101228:18)"
```

**Comment by apupier** [comment]
asked on StackOverflow if cookies can be used in Webview API: https://stackoverflow.com/questions/54041871/can-cookie-be-used-in-vs-code-webview-api

**Comment by apupier** [comment]
seems that cookie/localstorage cannot be activated anymore https://github.com/Microsoft/vscode/issues/48464#issuecomment-384105131 thanks @jaysensharma for the pointer)

**Comment by lhein** [comment]
How does that help us?

**Comment by apupier** [comment]
> How does that help us?

it explains the current errors and it is driving possible next steps:
- develop the feature in vs code
- find a way to have Atlasmap not using cookie/localstorzage
- use a different approach than Webview API

**Comment by apupier** [comment]
created https://github.com/apupier/vscode-atlasmap/tree/36-useWevViewAPI as a mor eup-to-date version with master

**Comment by lhein** [comment]
obsolete due to #76

**Pull request closed without merging**